core/src/main/java/org/apache/lucene/analysis/miscellaneous/DuplicateByteSequenceSpotter.java

```diff
@@ -0,0 +1,325 @@
+    // for speed.
+    class RootTreeNode extends TreeNode {
+
+        // A null-or-256 sized array that can be indexed into using a byte for
+        // fast access.
+        // Being near the root of the tree it is expected that this is a
+        // non-sparse array.
+        TreeNode[] children;
+
+        RootTreeNode(byte key, TreeNode parentNode, int depth) {
+            super(key, parentNode, depth);
+            bytesAllocated += ROOT_TREE_NODE_OBJECT_SIZE;
+        }
+
+        public TreeNode add(byte b, int depth) {
+            if (children == null) {
+                children = new TreeNode[256];
+                bytesAllocated += (RamUsageEstimator.NUM_BYTES_OBJECT_REF * 256);
+            }
+            int bIndex = b + 128;
+            TreeNode node = children[bIndex];
+            if (node == null) {
+                if (depth <= 1) {
+                    // Depths 0 and 1 use RootTreeNode impl and create
+                    // RootTreeNodeImpl children
+                    node = new RootTreeNode(b, this, depth);
+                } else {
+                    // Deeper-level nodes are less visited but more numerous
+                    // so use a more space-friendly data structure
+                    node = new LightweightTreeNode(b, this, depth);
+                }
+                children[bIndex] = node;
+            }
+            return node;
+        }
+
+        @Override
+        public short add(int byteSequence) {
+            throw new UnsupportedOperationException("Root nodes do not support byte sequences encoded as integers");
+        }
+
+    }
+
+    // Node implementation for use by the depth 3 branches of the tree that
+    // sacrifices speed for space.
+    final class LightweightTreeNode extends TreeNode {
+
+        // An array dynamically resized ranging from 0 to 256 values but
+        // frequently only sized 1 as most sequences leading to end leaves are
+        // one-off paths
+        ByteSequenceLeafNode[] children = null;
+
+        LightweightTreeNode(byte key, TreeNode parentNode, int depth) {
+            super(key, parentNode, depth);
+            bytesAllocated += LIGHTWEIGHT_TREE_NODE_OBJECT_SIZE;
+
+        }
+
+        @Override
+        public short add(int byteSequence) {
+            if (children == null) {
+                // Create array adding new child
+                children = new ByteSequenceLeafNode[1];
+                bytesAllocated += RamUsageEstimator.NUM_BYTES_ARRAY_HEADER;
+
+                ByteSequenceLeafNode child = new ByteSequenceLeafNode(byteSequence);
+                bytesAllocated += RamUsageEstimator.NUM_BYTES_OBJECT_REF;
+                children[0] = child;
+                return child.hitCount;
+            }
+            // Find existing child and increment count
+            for (ByteSequenceLeafNode child : children) {
+                if (child.byteSequence == byteSequence) {
```

Comment:
Makes sense to me. I think it is worth leaving a comment on the children array then.